### PR TITLE
Uploading large files works with the simulator.

### PR DESCRIPTION
**Pull request by bwbeach**

### CODE REVIEW

File: b2/raw_simulator.py [hunk 1]
```diff
@@ -229,18 +370,38 @@ def upload_file(
         url_match = re.match(r'https://upload.example.com/([^/]*)/([^/]*)', upload_url)
         if url_match is None:
             raise BadUploadUrl(upload_url)
+        if len(self.upload_errors) != 0:
```
Comment (ppolewicz) [inline b2/raw_simulator.py]: `if self.upload_errors:`

File: b2/raw_simulator.py [hunk 2]
```diff
@@ -229,18 +370,38 @@ def upload_file(
...
+        bucket = self._get_bucket_by_id(bucket_id)
+        return bucket.upload_file(
             upload_id, upload_auth_token, file_name, content_length, content_type, content_sha1,
             file_infos, data_stream
         )
 
+    def upload_part(
+        self, upload_url, upload_auth_token, part_number, content_length, sha1_sum, input_stream
+    ):
+        re.compile('https://upload.example.com/part/([^/]*)')
```
Comment (ppolewicz) [inline b2/raw_simulator.py]: `RE_UPLOAD_PART = re.compile('https://upload.example.com/part/([^/]*)')` in scope of the module and then:

File: b2/raw_simulator.py [hunk 3]
```diff
@@ -229,18 +370,38 @@ def upload_file(
...
+        url_match = re.match('https://upload.example.com/part/([^/]*)', upload_url)
```
Comment (ppolewicz) [inline b2/raw_simulator.py]: `url_match = RE_UPLOAD_PART.match(upload_url)`

File: b2/raw_simulator.py [hunk 4]
```diff
@@ -62,6 +78,39 @@ def as_list_files_dict(self):
...
+        last_part_number = max(part.part_number for part in self.parts if part is not None)
+        for part_number in six.moves.range(1, last_part_number + 1):
+            if self.parts[part_number] is None:
+                raise MissingPart(part_number)
+        my_part_sha1_array = [
+            self.parts[part_number].content_sha1
+            for part_number in six.moves.range(1, last_part_number + 1)
+        ]
+        if part_sha1_array != my_part_sha1_array:
+            raise
```
Comment (neube3) [inline b2/raw_simulator.py]: @bwbeach: What exception should be raised here?

**Pull request merged**